Initial app for office, department and personnel management.

ehaivala opened the pull request:
So far there's models, plugins and admin views for offices, departments and personnel. Tests and test data factories included as well.

Refs RND-5 & RND-6

ehaivala committed:
Package scaffolding

Basic packaging setup.

Refs RND-5 & RND-6

juhakujala commented on cmsplugin_offices_ai/templates/cmsplugin_offices_ai/department.html:
```diff
@@ -0,0 +1,18 @@
+<div class="department">
+    <h3>{{ department.name }}</h3>
+    <p class="description">{{ department.description }}</p>
```
Should there be `if department.description`?

juhakujala commented on cmsplugin_offices_ai/templates/cmsplugin_offices_ai/office.html:
```diff
@@ -0,0 +1,24 @@
+
+<div class="office">
+    <h3>{{ office.name }}</h3>
+    <p class="description">{{ office.description }}</p>
```
Should there be `if office.description`?

juhakujala commented on cmsplugin_offices_ai/models/personnel.py:
```diff
@@ -0,0 +1,89 @@
+        blank=True,
+        on_delete=models.SET_NULL,
+    )
+
```
You forgot about having a description text field for each personnel

juhakujala reviewed (changes requested):
Few things, other than that it looked good to me.
Not quite sure about the backend stuff, if there's something, but maybe someone else can take a better look on those...

ehaivala commented:
@juhakujala Fixed the things you pointed out.

juhakujala commented:
Fine by me, wait for some backend guys review though 😄

erej commented on cmsplugin_offices_ai/models/plugin_models.py:
```diff
@@ -0,0 +1,91 @@
+            return _('Department (%s)') % self.departments.first().name
+        return _('Departments (amount: %s)') % count
+
+    def copy_relations(self, old_instance):
```
Could there be one base class to have this kind of common functions? Now this function is just copy-pasted for every plugin

ehaivala replied:
Hmm yeah. These are kinda repetitive for sure.

ehaivala replied:
Actually there's no point in moving this to base class. We would still have the need to override `copy_relations` for each plugin model so we could call some helper function which basically does the same assignment with `getattr` and `setattr`. Yeah the comment in each is basically copy-pasted but I think it's good to have. Someone might not know Django CMS well enough to understand why the function is needed.

JsseL commented on cmsplugin_offices_ai/models/departments.py:
```diff
@@ -0,0 +1,43 @@
+    office = models.ForeignKey(
+        Office,
+        verbose_name=_('office'),
+        related_name='departments_in_office',
```
`office.departments_in_office` sounds quite verbose. Wouldn't `office.departments` have been enough?

JsseL commented on cmsplugin_offices_ai/models/departments.py:
```diff
@@ -0,0 +1,43 @@
+    )
+
+    # Display options
+    ordering = models.PositiveSmallIntegerField(
```
This has typically been called `order_no`. To me `ordering` references the type of ordering..

JsseL commented on cmsplugin_offices_ai/models/personnel.py:
```diff
@@ -0,0 +1,94 @@
+        on_delete=models.SET_NULL,
+    )
+
+    office_override = models.ForeignKey(
```
What's the use case for this?

ehaivala replied:
Person might belong to a department that belongs to different office the person actually works at. I figured this might be somewhat common special case in real world.

JsseL commented on cmsplugin_offices_ai/models/personnel.py:
```diff
@@ -0,0 +1,94 @@
+    )
+
+    # Display options
+    show_in_list = models.BooleanField(
```
This feels a bit hard to understand.. I would have no idea which lists this is related to. What is the original use case for this?

ehaivala replied:
It flags the person visible in plugins that list personnel. This allows person to be hidden from department list plugin personnel lists for example.

The help text was misleading because I flipped the logic at some point during development but left the help text unchanged. I've fixed it now.

JsseL commented on cmsplugin_offices_ai/models/personnel.py:
```diff
@@ -0,0 +1,94 @@
+        default=True,
+        help_text=_('Hides this person from personnel list plugins')
+    )
+    ordering = models.PositiveSmallIntegerField(
```
Again, I'd prefer `order_no`.

JsseL reviewed (commented):
Didn't quite understand the generic `InformationEntry` idea, but otherwise seems like good progress. :+1:

suutari-ai commented on .MANIFEST.in:
```diff
@@ -0,0 +1,4 @@
+include README.md
```
Python packages should have README in ReST format to include it as the long description for setup metadata.

suutari-ai commented on .coveragerc:
```diff
@@ -0,0 +1,7 @@
+    *management*
+    *migrations*
+    *site-packages*
+    *venv*
```
Move this stuff to setup.cfg section `[coverage:run]`.

ehaivala commented:
About `InformationEntry`:

Instead of creating fixed fields for each type of required information we can add the needed amount of entries for each person. And if we need more entry types at some point we can just add new value to `EntryType` enum instead of creating new fields for each model.

So basically the `InformationEntry` allows more flexible contact information because some people might have multiple phone numbers, emails or something else.

ehaivala commented:
Made some changes as requested:

* Fixed the verbose related name
* Changed `ordering` to `order_no`
* Converted `README.md` to `README.rst`
* Moved `.coveragerc` contents to `setup.cfg`

ehaivala committed:
Add initial models and admin views.

Add initial models and database migrations for them. Create basic django admin
views for the models.

Refs RND-5 & RND-6

ehaivala committed:
Add CMSPlugins and frontend templates

Add CMSPlugins and templates to display office, department and personnel lists.

Refs RND-5 & RND-6

ehaivala committed:
Add management command for creating dummy data

Add test data factories and management command utilizing them so there's quick
way to seed dummy data for testing and development purposes.

Refs RND-5 & RND-6

ehaivala committed:
Update README

Update README. Use reStructuredText instead of Markdown.

Refs RND-5 & RND-6

ehaivala committed:
Add tests and CI bits

Add simple test suite and CI bits

Refs RND-5 & RND-6

suutari-ai reviewed (approved):
Good work, @ehaivala!
